lib/runtime/storage/storagediff.go

```diff
@@ -0,0 +1,308 @@
+// states of a trie, such as those occurring during the execution of a block.
+// It records updates (upserts), deletions, and changes to child tries.
+// This mechanism facilitates applying state transitions efficiently.
+// Changes accumulated in storageDiff can be applied to a trie using
+// the `applyToTrie` method
+type storageDiff struct {
+	upserts        map[string][]byte
+	deletes        map[string]bool
+	childChangeSet map[string]*storageDiff
+	mtx            sync.RWMutex
+}
+
+// newChangeSet initialises and returns a new storageDiff instance
+func newStorageDiff() *storageDiff {
+	return &storageDiff{
+		upserts:        make(map[string][]byte),
+		deletes:        make(map[string]bool),
+		childChangeSet: make(map[string]*storageDiff),
+		mtx:            sync.RWMutex{},
+	}
+}
+
+// get retrieves the value associated with the key if it's present in the
+// change set and returns a boolean indicating if the key is marked for deletion
+func (cs *storageDiff) get(key string) ([]byte, bool) {
+	if cs == nil {
+		return nil, false
+	}
+
+	cs.mtx.RLock()
+	defer cs.mtx.RUnlock()
+
+	// Check in recent upserts if not found check if we want to delete it
+	if val, ok := cs.upserts[key]; ok {
+		return val, false
+	} else if deleted := cs.deletes[key]; deleted {
+		return nil, true
+	}
+
+	return nil, false
+}
+
+// upsert records a new value for the key, or updates an existing value.
+// If the key was previously marked for deletion, that deletion is undone
+func (cs *storageDiff) upsert(key string, value []byte) {
+	if cs == nil {
+		return
+	}
+
+	cs.mtx.Lock()
+	defer cs.mtx.Unlock()
+	// If we previously deleted this trie we have to undo that deletion
+	if cs.deletes[key] {
+		delete(cs.deletes, key)
+	}
+
+	cs.upserts[key] = value
+}
+
+// delete marks a key for deletion and removes it from upserts and
+// child changesets, if present.
+func (cs *storageDiff) delete(key string) {
+	if cs == nil {
+		return
+	}
+
+	cs.mtx.Lock()
+	defer cs.mtx.Unlock()
+
+	delete(cs.childChangeSet, key)
+	delete(cs.upserts, key)
+	cs.deletes[key] = true
+}
+
+// deleteChildLimit deletes lexicographical sorted keys from a child trie with
+// a maximum limit, potentially marking the entire child trie for deletion
+// if the limit is exceeded.
+// This method will do not consider keys created during actual block execution
+// https://spec.polkadot.network/chap-host-api#id-version-2-prototype-2
+func (cs *storageDiff) deleteChildLimit(keyToChild string,
+	currentChildKeys []string, limit int) (
+	deleted uint32, allDeleted bool) {
+	childChanges := cs.childChangeSet[keyToChild]
+	if childChanges == nil {
+		childChanges = newStorageDiff()
+	}
+
+	if limit == -1 {
+		cs.delete(keyToChild)
+		deletedKeys := len(childChanges.upserts) + len(currentChildKeys)
+		return uint32(deletedKeys), true
+	}
+
+	newKeys := maps.Keys(childChanges.upserts)
+	allKeys := append(newKeys, currentChildKeys...)
+	sort.Strings(allKeys)
```

Comment:
> // This method will do not consider keys created during actual block execution

using `childChanges.upserts` is the same as using keys created under actual block exec, right?

Reply:
Yep, `upserts` are the keys-values created during the actual block exec

Reply:
yes, but my question is: if we should not consider the keys created under actual block exec then why are we appending the `childChanges.upserts` to the `currentChildKeys` and then iterating over `allKeys` and deleting them from the `childChanges`? 

I see that you check if the current iterated key is in not the `newKeys` if so you decrease one from the limit however the if the key belongs to the `newKeys` then it will be deleted

Reply:
I updated the comment to `Keys created during the current block execution do not count toward the limit` as it said in the spec.

We are appending the keys because we have to delete them (even if they were created during the current block's exec) but we don't have to count them on the limit